src/contracts/GPv2Settlement.sol

```diff
@@ -262,17 +284,48 @@ contract GPv2Settlement {
         inTransfer.amount = inTransfer.amount.add(order.feeAmount);
     }
 
-    /// @dev Compute a unique identifier that represents a user order.
+    /// @dev Extracts specific order information from the standardized unique
+    /// order id of the protocol.
+    /// @param orderUid The unique identifier used to represent an order in
+    /// the protocol. This uid is the packed concatenation of the order digest,
+    /// the validTo order parameter and the address of the user who created the
+    /// order. It is used by the user to interface with the contract directly,
+    /// and not by calls that are triggered by the solvers.
+    /// @return orderDigest The unique digest associated to the parameters of an
+    /// order. See [`orderUidKey`] for details.
+    /// @return validTo The epoch time at which the order will stop being valid.
+    /// @return owner The address of the user who owns this order.
+    function extractOrderUidParams(bytes calldata orderUid)
+        internal
+        pure
+        returns (
+            bytes32 orderDigest,
+            uint32 validTo,
+            address owner
+        )
+    {
+        require(orderUid.length == 32 + 4 + 20, "GPv2: invalid uid");
+        // solhint-disable-next-line no-inline-assembly
```

Comment:
ubernit: Maybe a short comment as to why assembly is used? Something along the lines
```
Use assembly to decode packed data
```

Reply:
This is because `abi.decodePacked` doesn't exist, not because `abi.decode` is inefficient right?

Reply:
Definitely because `abi.decodePacked` doesn't exist. Something could have been hacked together with `abi.decode` and some bit operations, but it would have actually been less easy to read.

Reply:
The "hack" @fedgiac is referring to is writing `shr(z, calldataload(add(x.offset, y)))` either as:
- `uintN(abi.decode(x[y:], (uint256)) >> z)` if the packed data is near the front of the bytes or
- `uintN(abi.decode(x[y-(32+N):], (uint256)))` if its near the back of the bytes.
- If (bytes.length < 32) then it cannot be done!

I had suggested the assembly since, personally, I find it just as easy (if not easier) to read an understand (you don't have to explain why you read and shift vs read and mask in different cases for example, and instead reading each field works exactly the same). That being said I've been looking at assembly a lot lately so I may be biased :stuck_out_tongue:.